machida3/machida.pony

```diff
@@ -834,11 +834,48 @@ primitive Machida
 
   fun py_bytes_or_unicode_to_pony_string(p: Pointer[U8] box): String =>
     recover
-      let ps = @py_bytes_or_unicode_as_char(p)
-      let ps_size = @py_bytes_or_unicode_size(p)
-      String.copy_cpointer(ps, ps_size)
+      var ps: Pointer[U8] = Pointer[U8]
+      var ps_size: ISize = 0
+      var err: I32 = 0
+      if @py_bytes_check(p) == 0 then
+        ps = @PyUnicode_AsUTF8AndSize(p, addressof ps_size)
+      else
+        err = @PyBytes_AsStringAndSize(p, addressof ps,  addressof ps_size)
+      end
+      if (err != 0) or ps.is_null() then
+        @printf[U32]("unexpected non-string value\n".cstring())
+        Fail()
```

Comment:
Also, not sure if we should be using `Fail()` here or not. There are other parts of the code where `print_errors()` is used to print a stack trace if an error occurred but where we don't `Fail()` but instead raise an `error`. @aturley can you please let us know which is appropriate here (and in `py_bytes_or_unicode_to_pony_array`)?

Reply:
Good question as well. I'd like to make noise by default if we're not sure. Stack tracing here would help of course.